Forms: Add unread/read filtering

enejb opened the pull request:
This Pr Adds a unread/read filter to the dashboard. 
<img width="549" height="424" alt="Screenshot 2025-10-15 at 2 25 30 PM" src="https://github.com/user-attachments/assets/0f56eae0-7f9b-4980-b2cb-e9324546a80a" />


## Proposed changes:
* Adds filters for count and feedback endpoints. To only return data that is specific to the filter. 
* Update the UI to make it possible to filter. 
* Update the count store to use a key that matches the currentQuery so that the results are what you would expecte always.
* Invalidate the count store when feedback is marked and unmarked with the count. So that the count stays accurate.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to dashboard.
* Apply the unread filter. For both states ( unread and read ) and notice that you get the results that you would expect.
* Apply the filter for data and notice that it still works as expected. The same goes for source. 
* Clear all filters and notice that the count works as expected. 
* Mark things as read and unread and notice that the count looks like you would expected in the tabs.

enejb committed:
Update the endpoint to allow filtering by unread

enejb committed:
Update the counts endpoint to to return read and unread state

enejb committed:
Update the front end to show the filter and update the count.

enejb committed:
Update the count to always include a the current Query as key so that the filters always return the expected results.

enejb committed:
changelog

enejb committed:
Update the empty state to account for the filter.

enejb committed:
Invalidet the count store when the user marks things as unread so that the count is more accurate

lezama commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -301,32 +301,38 @@ public function register_routes() {
 						'validate_callback' => 'rest_validate_request_arg',
 					),
+					'is_unread' => array(
+						'description'       => 'Limit counts to read or unread feedback items.',
```
```suggestion
						'description'       => 'Limit results to read or unread feedback items.',
```

lezama commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -411,6 +418,12 @@ public function get_status_counts( $request ) {
 
+		// Filter by read/unread status using comment_status field.
+		if ( null !== $is_unread ) {
+			$comment_status     = rest_sanitize_boolean( $is_unread ) ? Feedback::STATUS_UNREAD : Feedback::STATUS_READ;
```
we already call `rest_sanitize_boolean` in the `sanitize_callback`?

enejb replied:
yeah this is not needed

lezama commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -803,6 +816,24 @@ public function prepare_item_for_response( $item, $request ) {
+		$args = parent::prepare_items_query( $args, $request );
+
+		if ( isset( $request['is_unread'] ) ) {
+			$is_unread              = rest_sanitize_boolean( $request['is_unread'] );
```
we already call `rest_sanitize_boolean` in the `sanitize_callback`?

enejb replied:
good call!

lezama commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -803,6 +816,24 @@ public function prepare_item_for_response( $item, $request ) {
+	 * @return array Modified query arguments.
+	 */
+	protected function prepare_items_query( $args = array(), $request = null ) {
+		$args = parent::prepare_items_query( $args, $request );
+
+		if ( isset( $request['is_unread'] ) ) {
+			$is_unread              = rest_sanitize_boolean( $request['is_unread'] );
+			$args['comment_status'] = $is_unread ? Feedback::STATUS_UNREAD : Feedback::STATUS_READ;
+		}
+
+		return $args;
+	}
```
😅

enejb replied:
This is not true. Since it works as expected. Nice try Copilot

enejb committed:
normalize values in cache key

enejb committed:
Invaliade even if only one of the requests succeeded

enejb committed:
Fix wording.

Co-authored-by: Miguel Lezama <miguel@automattic.com>

enejb committed:
don't double sanitize

enejb committed:
Fix: Keep track of the fields that are marked as read and need to be refetched.

CGastrell reviewed (dismissed):
We need to follow up to better invalidate/retain results through pagination and read/unread status update, but we can ship this so it doesn't get huge

enejb committed:
Make the js tests pass

enejb committed:
Fix pagination bug.

Currently when you are using the unread/read filter and you mark things as read the pagination breaks. Since the state on the client and the backend is not in sync any more. We fix this by passing in an array of invalid ids. This way the state between the browser and the backend matched and the pagination works as expected.

enejb committed:
Fix phan

enejb committed:
fix js tests

CGastrell reviewed (approved):
Works nicely! We'll follow up on any lingering glitches! :shipit:
